libflux: use common code for aux_set/aux_get

garlick opened the pull request:
This is a cleanup of the aux get/set functions in  the`flux_t`, `flux_mrpc_t`, `flux_future_t`, and `flux_message_t` objects to normalize their semantics, as called out in #1796.

I think this is a good idea as we stabilize our API, though  I'm less sure that the switch from a zhash to a one-off singly linked list is justified, though it should save some memory in those objects.

The header from the shared implementation summarizes the semantics:
```c
/* aux list - associate auxiliary data with a host object
 *
 * The object declares a 'struct aux_item *aux', initialized to NULL.
 * The object's destructor calls aux_destroy (&aux).
 * The object's aux_get/aux_set functions call aux_get ()/aux_set () below.
 *
 * An empty aux list is represented by a NULL pointer.
 *
 * It is legal to aux_set (key=NULL, value!=NULL).  aux_get (key=NULL) fails,
 * but aux_destroy () calls the anonymous value's destructor, if any.
 *
 * It is legal to aux_set () a duplicate key.  The new value replaces the old,
 * after calling its destructor, if any.
 *
 * It is legal to aux_set (key!=NULL, value=NULL).  Any value previously
 * stored under key is deleted, calling its destructor, if any.
 */
```

grondo commented:
> though I'm less sure that the switch from a zhash to a one-off singly linked list is justified, though it should save some memory in those objects.

Was there a memory issue that this was intended to fix, and if so a benchmark that shows that the performance and interface tradeoff is acceptable due to the memory usage savings? I guess I'd feel more comfortable if I knew where the sudden focus on memory usage above anything else is coming from (not that I'm arguing against the current approach, it just seemed like the change was coming from some experience someone had with usage.)

I've only quickly looked at the `struct aux_item` interface, but my first question was why the "aux list" interface was so unique that it now requires its own implementation of a linked list and a "different" interface. I also wonder if that choice of interface would make it more difficult to switch to a different underlying data structure in the future (maybe a leaner hash?), since the sort of "head of list" interface presented here is very list-specific.

Sorry about all the questions! I don't mean to sound like I'm dumping on this PR, but for my own benefit wanted to understand the choices made here.

(and BTW, *very* good idea to unify the "aux" interfaces across objects)

Finally, `libsubprocess` also has a similar interface `flux_subprocess_get_context(3)` and `flux_subprocess_set_context(3)`. I guess those need to be changed as well.

garlick commented:
Sorry, I didn't mean to get us off in the weeds talking about saving a few bytes of memory here and there.  The main justification of this PR is the interface normalization.  I did that by factoring out common code, and didn't mean to create a new and weird generic container.

I could change the implementation to use a zlist or a zhash and it wouldn't really matter.  My thought in doing it the way I did was that it was probably about the same amount of code, would have the side benefit of being lighter weight (yeah prob not justified), and wouldn't have the czmq dependency so it could be copied into flux-security if desired.  But after I got done, it was a little more complicated than I was comfortable with hence the hedging in the PR description about its justification.

chu11 commented on src/common/libflux/test/future.c:
```diff
@@ -10,7 +10,7 @@
 int aux_destroy_called;
 void *aux_destroy_arg;
-void aux_destroy (void *arg)
+void aux_destroy_fun (void *arg)
```
i don't know how to comment on a commit message, but noticed you have at typo ```wiht``` in the commit message for this change :-)

chu11 commented on src/common/libflux/mrpc.c:
```diff
@@ -526,29 +527,21 @@ flux_mrpc_t *flux_mrpc_pack (flux_t *h,
-    zhash_delete (mrpc->aux, name);
-    if (zhash_insert (mrpc->aux, name, aux) < 0) {
-        errno = ENOMEM;
+    if (!mrpc) {
```
don't know if it should be in this PR or perhaps we can just create a new issue for it.  But there are several other places in ```mrpc.c``` that do ```assert (mrpc->magic == MRPC_MAGIC)``` instead of ```if (!mrpc)```.  It should probably be consistent.

garlick replied:
I'd say that's a different issue.

garlick commented:
> Finally, libsubprocess also has a similar interface flux_subprocess_get_context(3) and flux_subprocess_set_context(3). I guess those need to be changed as well.
```c
/*
 *  Set arbitrary context `ctx` with name `name` on subprocess object `p`.
 *
 *  Returns 0 on success
 */
int flux_subprocess_set_context (flux_subprocess_t *p,
                                 const char *name, void *ctx);

/*
 *  Return pointer to any context associated with `p` under `name`. If
 *   no such context exists, then NULL is returned.
 */
void *flux_subprocess_get_context (flux_subprocess_t *p, const char *name);
```

Did you want to change these to:
```c
void *flux_subprocess_aux_get (flux_future_t *f, const char *name);

int flux_subprocess_aux_set (flux_future_t *f, const char *name,
                             void *aux, flux_free_f destroy);
```
or keep the names, but add a destructor to aux_set, and use the common implementation?

chu11 commented:
One extra comment:

```
 * It is legal to aux_set (key=NULL, value!=NULL).  aux_get (key=NULL) fails,
 * but aux_destroy () calls the anonymous value's destructor, if any.
```

I see this as a convenient optional way to stash something away for destruction later on.  But should this convenience be noted more explicitly in the manpage?  When seeing an ```aux_set```, function, I wouldn't think this behavior is allowed.

grondo commented:
Actually I did a quick test and, though the test is a bit contrived, this PR is actually kind of a big win. Here's the output from valgrind's massif of a run before this PR and one after. The test code creates 1024 futures using `flux_rpc`, then calls `flux_rpc_get` on each one before exiting (without calling `flux_future_destroy`)

Before pr #1797:
![massif-before](https://user-images.githubusercontent.com/741970/48018995-32412f00-e0e7-11e8-98bd-88a20260e82c.jpg)

After pr #1797:
![massif-after](https://user-images.githubusercontent.com/741970/48019022-46852c00-e0e7-11e8-9cae-22652ae2fe85.png)


This PR saves 4MiB,  or over 20% of heap memory on this testcase.

Just for my own interest, I may retry this testcase again with a reactor instead of synchronous gets. I'm wondering about the vast majority of heap memory which is coming from `msg_handler.c:fastpath_set`.

grondo commented:
> But should this convenience be noted more explicitly in the manpage? 

It is already documented in the `flux_future_create(3)` man page -- good idea that this should be copied to other manpages with `aux_set/get`

> Objects may be stored anonymously under a NULL _name_ to be scheduled for destruction without the option of retrieval.

grondo commented:
> Did you want to change these to:

@chu11 should probably comment, but I think your idea to develop a common interface applies also to libsubprocess, so I'd go with changing the function names.

chu11 commented:
> keep the names, but add a destructor to aux_set, and use the common implementation?

My vote would be to change to ```aux_set/aux_get```, add the destructor, and use common implementation.

garlick commented:
> I also wonder if that choice of interface would make it more difficult to switch to a different underlying data structure in the future (maybe a leaner hash?), since the sort of "head of list" interface presented here is very list-specific.

Although the parameter name `head` could have been better chosen not imply implementation details, I think the function prototypes where the "aux thing" is passed in by reference whenever it can be modified would work equally well for creating a container like zhash or zlist on first use.

The "empty list represented by NULL pointer" invariant was to avoid creating the container unconditionally, in case it wasn't used.  You could probably argue this is also premature optimization, although 3 of 4 implementations already included (duplicated) logic for this.

Would it be better if this was reimplemented in terms of a zhash?  Since there are now tests on the generic implementation, that should be quite easy.

garlick commented:
> This PR saves 4MiB, or over 20% of heap memory on this testcase.

WTF?!?

grondo commented:
> Although the parameter name head could have been better chosen not imply implementation details, I think the function prototypes where the "aux thing" is passed in by reference whenever it can be modified would work equally well for creating a container like zhash or zlist on first use.

Ok, that satisfies my concern on that count! (like I said I just quickly looked at the interface, sorry!)


> WTF?!?

Yes,  this simple testcase which basically creates 1024 futures seems to use 4MiB just in `zhash_new`, which is eliminated by your improvements. I'm totally convinced that your suspicion that zhash is overkill was completely correct. In fact, now I'm wondering about zhash in other places...

grondo commented:
> Would it be better if this was reimplemented in terms of a zhash? Since there are now tests on the generic implementation, that should be quite easy.

oh, no! I'm totally convinced now that zhash is not meant for use except when many hundereds of items are guaranteed to be stored, and certainly nowhere where it will be a short-lived data structure! I guess we get spoiled by the easy access to associative arrays as an interface in other languages, but have to be more careful in straight C.

Sorry that I caused so much noise in what should have been a simple PR.

garlick commented:
> Sorry that I caused so much noise in what should have been a simple PR.

Not at all, that was interesting.

grondo commented:
For the curious, here's a run of the same testcase using `flux_future_then` and the reactor instead of synchronous `flux_future_get`:

Before #1797:
![massif-before](https://user-images.githubusercontent.com/741970/48020531-f90abe00-e0ea-11e8-9e5d-737063e34d27.jpg)

After #1797:
![massif-after](https://user-images.githubusercontent.com/741970/48020549-032cbc80-e0eb-11e8-973a-b3bbff121508.jpg)

garlick commented:
> I'm totally convinced now that zhash is not meant for use except when many hundereds of items are guaranteed to be stored, and certainly nowhere where it will be a short-lived data structure! 

zhashx might be a little nicer, as I think it is more sophisticated about how it manages growth.

SteVwonder commented:
> zhashx might be a little nicer, as I think it is more sophisticated about how it manages growth.

Yeah, based on a quick glance at the source code, it looks like zhashx [is initialized](https://github.com/zeromq/czmq/blob/master/src/zhashx.c#L105) to 3 items to start (i.e., the first prime in [this prime list](https://github.com/zeromq/czmq/blob/master/src/zhash_primes.inc#L23)).

When expanding, it increases its "prime limit" by [5 primes](https://github.com/zeromq/czmq/blob/master/src/zhashx.c#L32).....and my knowledge of primes is nowhere good enough to estimate what that means in terms of a percent increase.  But it does start small, grow quickly, and then converge to approx doubling (which becomes really apparent when you look at the prime list which has the powers of two nicely labeled, and the approx gap between the powers of two is 5 primes):  `3 -> 17 -> 43 -> 103 -> 229 -> 487 ...`

chu11 commented:
> zhashx might be a little nicer, as I think it is more sophisticated about how it manages growth.

Which begs the question, how many places should we be using ```zhashx_t``` instead of ```zhash_t```?  Skimming code, I can see several places in the ```libsubprocess``` code that the change would probably be positive (number of channels and number of options in a command should be small on average).

grondo commented:
> Which begs the question, how many places should we be using zhashx_t instead of zhash_t? Skimming code, I can see several places in the libsubprocess code that the change would probably be positive (number of channels and number of options in a command should be small on average). 

Yeah, though it may be premature, we should probably open an issue, and optimize our search for sites where zhash is used in a data structure that might be rapidly used many times in a single address space -- e.g. `flux_future_t` [probably largely addressed in this pr], subprocess is another good one [before and after `flux-exec` max rss would be interesting], etc.)

garlick commented:
I added on some commits to
* check `flux_aux_set()` in a few more places
* make the proposed change to libsubprocess and update users
* try to normalize documentation for aux functions across objects
* added to unit test

I'll fix that commit typo when I eventually squash.  LMK if I should do that now and rebase on current master.

grondo commented:
Great, thanks! Since we have a couple smaller PRs going in first you can probably wait until those go in (#1805, #1798), but I'd be fine with a squash+rebase when you're ready.  Thanks for taking this one on. Good improvement!

garlick commented:
Rebased on master, fixed commit typo, squashed incremental development, and combined the API changes with updates to API users to avoid breaking git bisect.

grondo commented:
Great, thanks!
